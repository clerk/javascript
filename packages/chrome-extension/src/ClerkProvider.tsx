import Clerk from '@clerk/clerk-js';
import type { ClerkProp, ClerkProviderProps as ClerkReactProviderProps } from '@clerk/clerk-react';
import { __internal__setErrorThrowerOptions, ClerkProvider as ClerkReactProvider } from '@clerk/clerk-react';
import React from 'react';

import type { TokenCache } from './cache';
import { ChromeStorageCache } from './cache';
import { buildClerk } from './singleton';

__internal__setErrorThrowerOptions({
  packageName: '@clerk/chrome-extension',
});

type WebSSOClerkProviderCustomProps =
  | {
      syncSessionWithTab?: false;
      tokenCache?: never;
    }
  | {
      syncSessionWithTab: true;
      tokenCache?: TokenCache;
    };

type WebSSOClerkProviderProps = ClerkReactProviderProps & WebSSOClerkProviderCustomProps;

const WebSSOClerkProvider = (props: WebSSOClerkProviderProps): JSX.Element | null => {
  const { children, tokenCache: runtimeTokenCache, ...rest } = props;
  const { publishableKey = '' } = props;

  const [clerkInstance, setClerkInstance] = React.useState<ClerkProp>(null);

  // When syncSessionWithTab is set tokenCache is an optional parameter that defaults to ChromeStorageCache
  const tokenCache = runtimeTokenCache || ChromeStorageCache;

  React.useEffect(() => {
    (async () => {
      setClerkInstance(await buildClerk({ publishableKey, tokenCache }));
    })();
  }, []);

  if (!clerkInstance) {
    return null;
  }

  return (
    <ClerkReactProvider
      {...rest}
      Clerk={clerkInstance}
      standardBrowser={false}
    >
      {children}
    </ClerkReactProvider>
  );
};

const StandaloneClerkProvider = (props: ClerkReactProviderProps): JSX.Element => {
  const { children, ...rest } = props;

  return (
    <ClerkReactProvider
      {...rest}
      Clerk={Clerk}
    >
      {children}
    </ClerkReactProvider>
  );
};

type ChromeExtensionClerkProviderProps = WebSSOClerkProviderProps;

export function ClerkProvider(props: ChromeExtensionClerkProviderProps): JSX.Element | null {
  const { tokenCache, syncSessionWithTab, ...rest } = props;
  return syncSessionWithTab ? (
    <WebSSOClerkProvider
      {...props}
      tokenCache={tokenCache}
    />
  ) : (
    <StandaloneClerkProvider {...rest} />
  );
}
