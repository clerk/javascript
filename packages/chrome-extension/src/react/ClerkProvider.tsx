import type { Clerk } from '@clerk/clerk-js';
import type { ClerkProviderProps as ClerkReactProviderProps } from '@clerk/clerk-react';
import { ClerkProvider as ClerkReactProvider } from '@clerk/clerk-react';
import React from 'react';

import { createClerkClient } from '../internal/clerk';
import type { StorageCache } from '../internal/utils/storage';

type SyncHost = `${'http' | 'https'}://${string}`;
type ChromeExtensionClerkProviderProps = ClerkReactProviderProps & {
  storageCache?: StorageCache;
  syncHost?: SyncHost;
};

export function ClerkProvider(props: ChromeExtensionClerkProviderProps): JSX.Element | null {
  const { children, storageCache, syncHost, ...rest } = props;
  const { publishableKey = '' } = props;

  const [clerkInstance, setClerkInstance] = React.useState<Clerk | null>(null);

  React.useEffect(() => {
    void (async () => {
      setClerkInstance(await createClerkClient({ publishableKey, storageCache, syncHost }));
    })();
  }, [publishableKey, storageCache, syncHost]);

  if (!clerkInstance) {
    return null;
  }

  return (
    <ClerkReactProvider
      {...rest}
      Clerk={clerkInstance}
      standardBrowser={!syncHost}
    >
      {children}
    </ClerkReactProvider>
  );
}
