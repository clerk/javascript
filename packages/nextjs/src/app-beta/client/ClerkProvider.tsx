'use client';
// !!! Note the import from react
import type { ClerkProviderProps } from '@clerk/clerk-react';
import { ClerkProvider as ReactClerkProvider } from '@clerk/clerk-react';
import { usePathname, useRouter } from 'next/navigation';
import React, { useCallback, useEffect } from 'react';

declare global {
  export interface Window {
    __clerk_nav_await: Array<(value: void) => void>;
    __clerk_nav: (to: string) => Promise<void>;
  }
}

export const useAwaitableNavigate = () => {
  // eslint-disable-next-line @typescript-eslint/unbound-method
  const { push, refresh } = useRouter();
  const pathname = usePathname();

  useEffect(() => {
    window.__clerk_nav = (to: string) => {
      return new Promise(res => {
        window.__clerk_nav_await.push(res);
        if (to === pathname) {
          refresh();
        } else {
          push(to);
        }
      });
    };
  }, [pathname]);

  useEffect(() => {
    if (typeof window.__clerk_nav_await === 'undefined') {
      window.__clerk_nav_await = [];
    }
    window.__clerk_nav_await.forEach(resolve => resolve());
    window.__clerk_nav_await = [];
  });

  return useCallback((to: string) => {
    return window.__clerk_nav(to);
  }, []);
};

export function ClerkProvider(props: ClerkProviderProps) {
  const navigate = useAwaitableNavigate();
  return (
    <ReactClerkProvider
      navigate={navigate}
      {...props}
    />
  );
}
