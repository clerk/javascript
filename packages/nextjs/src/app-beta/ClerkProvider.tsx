import type { IsomorphicClerkOptions } from '@clerk/clerk-react';
import type { InitialState, PublishableKeyOrFrontendApi } from '@clerk/types';
import React from 'react';

import { initialState } from './auth';
import { ClerkProvider as ClerkProviderClient } from './client/ClerkProvider';

type NextAppClerkProviderProps = {
  children: React.ReactNode;
} & Omit<IsomorphicClerkOptions, keyof PublishableKeyOrFrontendApi> &
  Partial<PublishableKeyOrFrontendApi>;

export function ClerkProvider(props: NextAppClerkProviderProps) {
  const state = (initialState()?.__clerk_ssr_state || { sessionId: null, orgId: null, userId: null }) as InitialState;
  return (
    // @ts-expect-error
    <ClerkProviderClient
      frontendApi={process.env.NEXT_PUBLIC_CLERK_FRONTEND_API || ''}
      publishableKey={process.env.NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY || ''}
      proxyUrl={process.env.NEXT_PUBLIC_CLERK_PROXY_URL}
      initialState={state}
      {...props}
    />
  );
}
