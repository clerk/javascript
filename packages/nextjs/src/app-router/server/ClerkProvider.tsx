import type { IsomorphicClerkOptions } from '@clerk/clerk-react';
import type { InitialState, PublishableKeyOrFrontendApi } from '@clerk/types';
import React from 'react';

import { mergeNextClerkPropsWithEnv } from '../../utils/mergeNextClerkPropsWithEnv';
import { ClientClerkProvider } from '../client/ClerkProvider';
import { initialState } from './auth';

type NextAppClerkProviderProps = React.PropsWithChildren<
  Omit<IsomorphicClerkOptions, keyof PublishableKeyOrFrontendApi> & Partial<PublishableKeyOrFrontendApi>
>;

export function ClerkProvider(props: NextAppClerkProviderProps) {
  const { children, ...rest } = props;
  const state = initialState()?.__clerk_ssr_state as InitialState;

  return (
    <ClientClerkProvider
      {...mergeNextClerkPropsWithEnv(rest)}
      // @ts-ignore
      initialState={state}
    >
      {children}
    </ClientClerkProvider>
  );
}
