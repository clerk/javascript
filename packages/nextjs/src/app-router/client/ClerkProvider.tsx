'use client';
import { ClerkProvider as ReactClerkProvider } from '@clerk/clerk-react';
import { useRouter } from 'next/navigation';
import React from 'react';

import { ClerkNextOptionsProvider } from '../../client-boundary/NextOptionsContext';
import { useSafeLayoutEffect } from '../../client-boundary/useSafeLayoutEffect';
import type { NextClerkProviderProps } from '../../types';
import { useAwaitableNavigate } from './useAwaitableNavigate';

declare global {
  export interface Window {
    __clerk_nav_await: Array<(value: void) => void>;
    __clerk_nav: (to: string) => Promise<void>;
  }
}

export const ClientClerkProvider = (props: NextClerkProviderProps) => {
  const { __unstable_invokeMiddlewareOnAuthStateChange = true } = props;
  const router = useRouter();
  const navigate = useAwaitableNavigate();

  useSafeLayoutEffect(() => {
    window.__unstable__onBeforeSetActive = () => {
      if (__unstable_invokeMiddlewareOnAuthStateChange) {
        router.refresh();
        router.push(window.location.href);
      }
    };

    window.__unstable__onAfterSetActive = () => {
      router.refresh();
    };
  }, []);

  const mergedProps = { ...props, navigate };
  return (
    <ClerkNextOptionsProvider options={mergedProps}>
      {/*// @ts-ignore*/}
      <ReactClerkProvider {...mergedProps} />
    </ClerkNextOptionsProvider>
  );
};
