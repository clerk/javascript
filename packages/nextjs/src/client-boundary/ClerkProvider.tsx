'use client';

import React from 'react';

import { ClerkProvider as PageClerkProvider } from '../pages/ClerkProvider';
import { type NextClerkProviderProps } from '../types';

/**
 * This is a compatibility layer to support a single ClerkProvider component in both the app and pages routers.
 * It also needs to support both Next.js before and after v13, hence the dynamic imports for certain modules.
 */
export function ClerkProvider(props: NextClerkProviderProps) {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { useRouter } = require('next/compat/router');
    const router = useRouter();

    return router ? <PageClerkProvider {...props} /> : <AppClerkProvider {...props} />;
  } catch (error) {
    // Silently ignore the error
    return <PageClerkProvider {...props} />;
  }
}

function AppClerkProvider(props: NextClerkProviderProps) {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const { ClientClerkProvider } = require('../app-router/client/ClerkProvider');

  return <ClientClerkProvider {...props} />;
}
