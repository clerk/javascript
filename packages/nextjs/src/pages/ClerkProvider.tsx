import { __internal__setErrorThrowerOptions, ClerkProvider as ReactClerkProvider } from '@clerk/clerk-react';
import { useRouter } from 'next/router';
import React from 'react';

import { ClerkNextOptionsProvider } from '../client-boundary/NextOptionsContext';
import { useSafeLayoutEffect } from '../client-boundary/useSafeLayoutEffect';
import type { NextClerkProviderProps } from '../types';
import { invalidateNextRouterCache } from '../utils/invalidateNextRouterCache';
import { mergeNextClerkPropsWithEnv } from '../utils/mergeNextClerkPropsWithEnv';

__internal__setErrorThrowerOptions({ packageName: '@clerk/nextjs' });

export function ClerkProvider({ children, ...props }: NextClerkProviderProps): JSX.Element {
  const { __unstable_invokeMiddlewareOnAuthStateChange = true } = props;
  const { push } = useRouter();
  ReactClerkProvider.displayName = 'ReactClerkProvider';

  useSafeLayoutEffect(() => {
    window.__unstable__onBeforeSetActive = invalidateNextRouterCache;
  }, []);

  useSafeLayoutEffect(() => {
    window.__unstable__onAfterSetActive = () => {
      // Re-run the middleware every time there auth state changes.
      // This enables complete control from a centralised place (NextJS middleware),
      // as we will invoke it every time the client-side auth state changes, eg: signing-out, switching orgs, etc.\
      if (__unstable_invokeMiddlewareOnAuthStateChange) {
        void push(window.location.href);
      }
    };
  }, []);

  const navigate = (to: string) => push(to);
  const mergedProps = mergeNextClerkPropsWithEnv({ ...props, navigate });
  // withServerSideAuth automatically injects __clerk_ssr_state
  // getAuth returns a user-facing authServerSideProps that hides __clerk_ssr_state
  // @ts-expect-error initialState is hidden from the types as it's a private prop
  const initialState = props.authServerSideProps?.__clerk_ssr_state || props.__clerk_ssr_state;

  return (
    <ClerkNextOptionsProvider options={mergedProps}>
      {/*@ts-expect-error*/}
      <ReactClerkProvider
        {...mergedProps}
        initialState={initialState}
      >
        {children}
      </ReactClerkProvider>
    </ClerkNextOptionsProvider>
  );
}
