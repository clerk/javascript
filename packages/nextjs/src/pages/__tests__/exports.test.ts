import * as publicExports from '../ClerkProvider';

describe('/client public exports', () => {
  it('should not include a breaking change', () => {
    expect(publicExports).toMatchSnapshot();
  });
});
