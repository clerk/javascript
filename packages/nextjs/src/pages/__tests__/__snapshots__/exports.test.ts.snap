// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/client public exports should not include a breaking change 1`] = `
{
  "ClerkProvider": [Function],
}
`;
